daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,163 @@
+  def name: String
+  override def walkDSOM[T](walker: AbstractDSOMWalker[T]): Unit = {
+    walker.onTermBegin(this)
+    if (isComplexType) {
+      walker.onTypeBegin(Right(complexType))
+      complexType.group.walkDSOM(walker)
+      walker.onTypeBegin(Right(complexType))
+    } else {
+      walker.onTypeBegin(Left(simpleType))
+      walker.onTypeBegin(Left(simpleType))
+    }
+    walker.onTermEnd(this)
+  }
+
+}
+
+trait RootView extends ElementBaseView
+
+/**
+ * A class designed to walk the internal representation of a DFDL Schema File.
+ *
+ * There are 2 main event handlers an implementing class has to worry about: one for Terms, and
+ * another for Types.  These are called as the DSOM is walked.
+ *
+ * Though recursion is used here to define the walk, it is not advised to use recursion between
+ * these event handlers.  Instead, consider a stack-like structure, as the DFDL Schema structure
+ * as well as the recursive method call structure can be represented by trees.
+ * @tparam T the return type of walkDSOMSchema and onWalkEnd.  This should be a class used to
+ *           represent the structure of the Schema.  You could always make it Unit if you don't
+ *           want these methods to return anything.
+ */
+abstract class AbstractDSOMWalker[T] {
+
+  /**
+   * Method to be called on the beginning of the traversal.  It is recommended to add some
+   * sort of wrapper element to a stack if you're doing a typical stack-based traversal.
+   *
+   * @param root the root element of the DFDL Schema
+   */
+  protected def onWalkBegin(root: RootView): Unit
+
+  /**
+   * Method to be called when the traversal concludes.  It is recommended to put any post-processing
+   * and anything to tidy up the stack or the result here.
+   *
+   * @param root the root element of the DFDL Schema
+   * @return the result of the traversal
+   */
+  protected def onWalkEnd(root: RootView): T
+
+  /**
+   * Method to be called whenever any element that is a Term is encountered.
+   * This applies to Sequence, Choice, GroupRef, ElementBase, etc.
+   *
+   * It is highly recommended that, when implementing this method, you pattern match
+   * some of these different sub-types at some point to handle each accordingly.
+   *
+   * @param termElement the term element
+   */
+  def onTermBegin(termElement: TermView): Unit
+
+  /**
+   * Method to be called when a Term element has finished processing
+   *
+   * @param termElement the term element
+   */
+  def onTermEnd(termElement: TermView): Unit
+
+  /**
+   * Method to be called when either a Simple or Complex Type element has been encountered.
+   * This is just the element for <simpleType> or <complexType>, so no implementation is necessary if you
+   * do not care about these wrapper portion; its children will be called automatically.
+   *
+   * @param typeElement either a Simple or Complex type
+   */
+  def onTypeBegin(typeElement: Either[SimpleTypeView, ComplexTypeView]): Unit
```

Comment:
Either's are a scala construct, which means this would likely be difficult, or at least uncomfortable, for a Java user to use. Perhaps we have a shared parent of SimpleTypeView and ComplexTypeView and require the user to use isntanceOf, similar to the onTerm methods?

Reply:
I opted for creating a shared parent in the latest revision - if we decide to stick with that then yes, in a Java implementation they would have to do instanceOf and casting